[CI] Test Failure in org.elasticsearch.client.BulkProcessorRetryIT#testBulkRejectionLoadWithBackoff

This is a result of #40866 

We're running into https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+intake/1107/console

```
  2> REPRODUCE WITH: ./gradlew :server:integTest --tests "org.elasticsearch.action.bulk.BulkProcessorRetryIT.testBulkRejectionLoadWithBackoff" -Dtests.seed=A9EF087F2D4CFAB -Dtests.security.manager=true -Dtests.locale=sr-Latn-RS -Dtests.timezone=America/Martinique -Dcompiler.java=12 -Druntime.java=8
  2> java.lang.AssertionError: Unexpected failure
        at __randomizedtesting.SeedInfo.seed([A9EF087F2D4CFAB:5EF9783BACF11D86]:0)
        at org.elasticsearch.action.bulk.BulkProcessorRetryIT.executeBulkRejectionLoad(BulkProcessorRetryIT.java:143)
        at org.elasticsearch.action.bulk.BulkProcessorRetryIT.testBulkRejectionLoadWithBackoff(BulkProcessorRetryIT.java:69)

        Caused by:
        RemoteTransportException[[node_s0][127.0.0.1:40957][indices:data/write/bulk]]; nested: EsRejectedExecutionException[rejected execution of org.elasticsearch.transport.InboundHandler$RequestHandler@61f4536b on EsThreadPoolExecutor[name = node_s0/write, queue capacity = 30, org.elasticsearch.common.util.concurrent.EsThreadPoolExecutor@1494a6e[Running, pool size = 1, active threads = 1, queued tasks = 30, completed tasks = 525]]];

            Caused by:
            org.elasticsearch.common.util.concurrent.EsRejectedExecutionException: rejected execution of org.elasticsearch.transport.InboundHandler$RequestHandler@61f4536b on EsThreadPoolExecutor[name = node_s0/write, queue capacity = 30, 
```

The problem here is that by moving the bulk requests' initial handling to the WRITE pool entirely we lost the bulk processor's retry functionality for the case of the initial request getting rejected outright.
We will need to discuss whether to find a way to retry this after-all or adjust the test I think.

--- a/client/rest-high-level/src/test/java/org/elasticsearch/client/BulkProcessorRetryIT.java
+++ b/client/rest-high-level/src/test/java/org/elasticsearch/client/BulkProcessorRetryIT.java
@@ -29,6 +29,7 @@ import org.elasticsearch.action.index.IndexRequest;
 import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.common.xcontent.XContentType;
 import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.transport.RemoteTransportException;
 
 import java.util.Collections;
 import java.util.Iterator;
@@ -56,7 +57,6 @@ public class BulkProcessorRetryIT extends ESRestHighLevelClientTestCase {
         executeBulkRejectionLoad(BackoffPolicy.noBackoff(), rejectedExecutionExpected);
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/41324")
     public void testBulkRejectionLoadWithBackoff() throws Throwable {
         boolean rejectedExecutionExpected = false;
         executeBulkRejectionLoad(BackoffPolicy.exponentialBackoff(), rejectedExecutionExpected);
@@ -122,9 +122,14 @@ public class BulkProcessorRetryIT extends ESRestHighLevelClientTestCase {
                     }
                 }
             } else {
-                Throwable t = (Throwable) response;
-                // we're not expecting any other errors
-                throw new AssertionError("Unexpected failure", t);
+                if (response instanceof RemoteTransportException
+                    && ((RemoteTransportException) response).status() == RestStatus.TOO_MANY_REQUESTS && rejectedExecutionExpected) {
+                    // ignored, we exceeded the write queue size with dispatching the initial bulk request
+                } else {
+                    Throwable t = (Throwable) response;
+                    // we're not expecting any other errors
+                    throw new AssertionError("Unexpected failure", t);
+                }
             }
         }
 

--- a/server/src/main/java/org/elasticsearch/action/bulk/Retry.java
+++ b/server/src/main/java/org/elasticsearch/action/bulk/Retry.java
@@ -26,6 +26,7 @@ import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.rest.RestStatus;
 import org.elasticsearch.threadpool.Scheduler;
 import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.RemoteTransportException;
 
 import java.util.ArrayList;
 import java.util.Iterator;
@@ -118,11 +119,15 @@ public class Retry {
 
         @Override
         public void onFailure(Exception e) {
-            try {
-                listener.onFailure(e);
-            } finally {
-                if (retryCancellable != null) {
-                    retryCancellable.cancel();
+            if (e instanceof RemoteTransportException && ((RemoteTransportException) e).status() == RETRY_STATUS && backoff.hasNext()) {
+                retry(currentBulkRequest);
+            } else {
+                try {
+                    listener.onFailure(e);
+                } finally {
+                    if (retryCancellable != null) {
+                        retryCancellable.cancel();
+                    }
                 }
             }
         }

--- a/server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java
+++ b/server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java
@@ -64,7 +64,6 @@ public class BulkProcessorRetryIT extends ESIntegTestCase {
         executeBulkRejectionLoad(BackoffPolicy.noBackoff(), rejectedExecutionExpected);
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/41324")
     public void testBulkRejectionLoadWithBackoff() throws Throwable {
         boolean rejectedExecutionExpected = false;
         executeBulkRejectionLoad(BackoffPolicy.exponentialBackoff(), rejectedExecutionExpected);

